rfcs/react-components/convergence/motion-definition-n-apis.md

```diff
@@ -0,0 +1,1079 @@
+
+<details>
+<summary>Why factories over hooks?</summary>
+
+To execute animations, we need a motion definition to create an animation. With hooks, this definition can be defined also inside a component, as illustrated below:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+function MyComponent() {
+  const motionRef = useAtomMotion({
+    keyframes: [{ opacity: 0 }, { opacity: 1 }],
+    duration: 2000,
+  });
+
+  return <div ref={motionRef}>Hello world!</div>;
+}
+```
+
+This approach presents a design issue: on every render, a new definition object is created and passed to the hook. At the same time, the motion must be a dependency to start the animation. Consequently, a new animation is created and started on every render, as depicted in the example below:
+
+```tsx
+function useAtomMotion(motion) {
+  const elementRef = React.useRef();
+
+  React.useEffect(() => {
+    const animation = elementRef.current.animate(motion.keyframes, {
+      duration: motion.duration,
+    });
+
+    return () => {
+      animation.cancel();
+    };
+  }, [motion]);
+}
+```
+
+With the factories pattern, the same problem is not possible, as a motion is defined outside a component, as shown below:
+
+```tsx
+function createAtomMotion(motion) {
+  return React.forwardRef((props, ref) => {
+    const elementRef = React.useRef();
+
+    React.useEffect(() => {
+      elementRef.current = elementRef.current.animate(motion.keyframes, {
+        duration: motion.duration,
+      });
+
+      return () => {
+        elementRef.current.cancel();
+      };
+      // `motion` does not need to be a dependency as it's defined outside the component
+    }, []);
+
+    return React.cloneElement(props.children, { ref: elementRef });
+  });
+}
+```
+
+</details>
+
+### Motion definition
+
+As the Web Animations API is framework-agnostic, we don't need to use Griffel to generate CSS keyframes. Instead, we can define them in JS:
+
+```tsx
+const fadeEnterSlow: MotionAtom = {
+  keyframes: [{ opacity: 0 }, { opacity: 1 }],
+
+  duration: 500,
+};
+```
+
+<details>
+<summary>What about CSS variables & tokens?</summary>
+
+The proposal suggests using JavaScript-based tokens instead of CSS variables, for example:
+
+```tsx
+const fadeEnterSlow: MotionAtom = {
+  keyframes: [{ opacity: 0 }, { opacity: 1 }],
+
+  duration: motionTokens.durationSlow,
+};
+```
+
+In the same time, CSS variables can be used inside `keyframes` directly:
+
+```tsx
+import { tokens } from '@fluentui/react-components';
+
+const flash: MotionAtom = {
+  keyframes: [
+    { backgroundColor: 'white' },
+    // 💡`tokens.colorBrandBackground` is a CSS variable i.e. `var(--colorBrandBackground)`
+    { backgroundColor: tokens.colorBrandBackground },
+    { backgroundColor: 'white' },
+  ],
+
+  duration: 500,
```

Comment:
Can we think of a way to specify `duration` and `easing` for each stage of an animation?

One example is the Drawer and Dialog components, which have different easing curves for each stage (entering or exiting). 
For presence-like animations, it is also quite useful to have "entering" faster than "exiting" animations.

Reply:
@marcosmoura that's already handled by design. `MotionAtom` is an atomic unit while, `PresenceMotion` is a compound from `enter` and `exit` atoms. So the described case, we can do:

```ts
const dialogPresence: PresenceMotion = {
  enter: {
    keyframes: [],
    duration: 500,
  },
  exit: {
    keyframes: [],
    duration: 300,
  }
}
```

However, this made me realize that `PresenceMotion` type is not explained in the RFC, added a note in c79b772.

Reply:
Maybe I was not completely clear here.
Elements can have multiple animations with different duration and/or easing configurations:
`animation: fade 1s ease-out, scale 2s ease-in`.

Also, an element can have different multiple animations depending on whether is opening/closing, for example:

``` css
.open {
	animation: open-fade 1s ease-in, open-scale 2s ease-in-out;
}

.close {
	animation: close-fade 1.5s ease-out, close-scale 2.5s ease-out;
}
```

How can we replicate that using the proposed API?


I can propose a way that would be the `createAtom` also accepting an array of motion definitions, instead of only being able to receive a single definition. A complex scenario would be:

``` ts
const fade: MotionAtom = {
  keyframes: [{ opacity: 0 }, { opacity: 1 }],
  duration: 500,
}

const scale: MotionAtom = {
  keyframes: [
    { transform: "scale(0)" }, 
    { transform: "scale(0.7)", offset: 0.5 }, 
    { transform: "scale(1)" }
  ],
  duration: 800,
}

const FadeScale = createAtom([scale, fade])
```

Internally, it would call the `animate` method multiple times, for each array item.

Another alternative, could be introducing a new method for that, something like:
``` ts
const fade: MotionAtom = {
  keyframes: [{ opacity: 0 }, { opacity: 1 }],
  duration: 500,
}

const scale: MotionAtom = {
  keyframes: [{ transform: "scale(0)" }, { transform: "scale(1)" }],
  duration: 800,
}

/* Single */
const Fade = createAtom(fade)
const Scale = createAtom(scale)

/* Multiple */
const FadeScale = createAtoms([scale, fade])
```

Please note that the `createAtoms` name is just to differentiate. It can be anything else that makes more sense.

Reply:
https://github.com/layershifter/office-ui-fabric-react/blob/docs/motion-rfc/rfcs/react-components/convergence/motion-definition-n-apis.md#motion-sequencing--grouping

@marcosmoura this is out of scope as I have not seen usages in our codebase or requests to implement this yet. It's not a big problem to implement, this have not been raised as a requirement yet.